File: Data Structures & Algorithms/longest-increasing-subsequence/submission-10.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        memo = {}
        def dfs(p: int, i: int) -> int:
            if i == n:
                return 0
            if (p, i) in memo:
                return memo[(p, i)]

            lis = dfs(p, i + 1)
            if p == -1 or nums[i] > nums[p]:
                lis = max(lis, 1 + dfs(i, i + 1))

            memo[(p, i)] = lis
            return lis

        return dfs(-1, 0)




# nums = [9,1,4,2,3,3,7]

#                         O
#             9                   []
        
    

File: Data Structures & Algorithms/subsets/submission-10.cpp
class Solution {
private:
    void dfs(int i, vector<int>& nums, vector<int>& current, vector<vector<int>>& result) {
        if (i == nums.size()) {
            result.push_back(current);
            return;
        }

        current.push_back(nums[i]);
        dfs(i + 1, nums, current, result);
        current.pop_back();
        dfs(i + 1, nums, current, result);
    }

public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> current;
        dfs(0, nums, current, result);
        return result;
    }
};



    //             (0)
    //     (1)             (1)
    // (2)     (2)     (2)     (2)